### PR TITLE
Added job deletion feature and replaceVM capability for prealloc.

### PR DESCRIPTION
**Pull request by icanberk**

### CODE REVIEW

File: jobQueue.py [hunk 1]
```diff
@@ -213,21 +212,30 @@ def unassignJob(self, job):
...
 		status = -1
 		if id in self.jobQueue:
 			status = 0
 			job = self.jobQueue[id]
-			del self.jobQueue[id]
 			if job.trace is None:
 				job.trace = []
-			job.trace.append("%s|%s" %  (time.ctime(time.time()+time.timezone), reason))
+				job.trace.append("%s|%s" %  (time.ctime(time.time()+time.timezone), reason))
+			if job.assigned = False:
```
Comment (ymzong) [inline jobQueue.py]: `job.assigned == False`, or `not job.trace`?

**Comment by icanberk** [comment]
@sinancepel you need to update this change with the current conventions.

We don't directly delete anything from queue, change attributes of job, or append traces. There are class methods for doing all of this.

**Comment by mihirpandya** [comment]
@sinancepel Any update on this? This PR has been around for long enough!

**Comment by mihirpandya** [comment]
@sinancepel Can you test this with docker? Also, note that `jobQueue` is `liveJobs` in `jobQueue.py`. Also found a bunch of merge conflicts in `tangoObjects.py`, `jobQueue.py` and `preallocator.py`.

File: preallocator.py [hunk 2]
```diff
@@ -237,8 +263,12 @@ def getPool(self, vmName):
             machine[1].put(vm)
             self.machines.set(vmName, machine)
         self.lock.release()
-        
+
+<<<<<<< HEAD
         result["pool"] = self.machines.get(vmName)[0]
+=======
+        result["pool"] = self.machines[vmName][0]
+>>>>>>> 7e264098d81a28c61a00148bd40b902b20070454
```
Comment (mihirpandya) [inline preallocator.py]: @sinancepel Bro, fix your merge conflicts. Did you even run your code?

File: preallocator.py [hunk 3]
```diff
@@ -187,8 +209,12 @@ def createVM(self, vm):
...
         self.addVM(newVM)
         self.freeVM(newVM)
-        self.log.debug("createVM: Added vm %s to pool %s" % (newVM.id, newVM.name))  
-
+        self.log.debug("createVM: Added vm %s to pool %s" % (newVM.id, newVM.name))
+    def restartVM(self, vmName, id):
+        #destroy vm
+        #add vm with same id
+        #free vm
+        pass
```
Comment (mihirpandya) [inline preallocator.py]: Is this yet to be implemented?

File: jobQueue.py [hunk 4]
```diff
@@ -261,7 +261,13 @@ def makeDead(self, id, reason):
             self.log.info("makeDead| Job is in the queue")
             status = 0
             job = self.jobQueue.get(id)
-            self.jobQueue.delete(id)
+            if job.assigned:
+                #If the job is assigned, kill the VM, and create a new one.
+                newVM = preallocator.replaceVM()
+                self.jobQueue.delete(id)
```
Comment (mihirpandya) [inline jobQueue.py]: We are no longer using `jobQueue. Instead, it's called `liveQueue`.

Comment (icanberk) [inline jobQueue.py]: Actually `liveJobs`

Comment (mihirpandya) [inline jobQueue.py]: You get me bro <3

Comment (icanberk) [inline jobQueue.py]: I feel ya.

**Comment by mihirpandya** [comment]
The worker thread always destroys the VM after a job is completed. Try simply adding one job and seeing what happens after that.

**Comment by nayak16** [comment]
@yrkumar was this fixed? Seems like this was long before my time

**Comment by nayak16** [comment]
Integrating these changes with new conventions and updates is causing too many runtime issues and conflicts. Closing PR and writing my own delete branch feature using aspects from this branch.

**Pull request closed without merging**